src/Microsoft.DotNet.Cli.Utils/EnvironmentProvider.cs

```diff
@@ -141,5 +141,14 @@ public bool GetEnvironmentVariableAsBool(string name, bool defaultValue)
             }
         }
 
+        public string GetEnvironmentVariable(string variable, EnvironmentVariableTarget target)
+        {
+            return Environment.GetEnvironmentVariable(variable, target);
```

Comment:
necessary? It is just a redirection

Note, EnvironmentVariableTarget is netcoreapp2.0 feature. It cannot be added to PlatformAbstraction (netstandard).

Reply:
It's necessary for the calls made by `WindowsEnvironmentPath`; otherwise we can't mock them in the tests.

`Environment.GetEnvironmentVariable(string, EnvironmentVariableTarget)` is netstandard 2.0: https://raw.githubusercontent.com/dotnet/standard/master/docs/versions/netstandard2.0_ref.md

`Microsoft.DotNet.Cli.Utils` is netstandard 2.0 and this builds without issue.

Reply:
Just general discussion, this specific change is good:

The downside of mock is it gives you false sense of security. It is ok for API like this (long existed, well know API). 

A good example is: What happens when there is no such env available? Is it returning a null, empty string or just throw? If you assume just return empty and you code your mock like it. You get a 100% code coverage, and you see there are tests. But you will get a surprise when it actually returns a null.

Also, what will this API behave on mac? There is no such concept of EnvironmentVariableTarget.Machine.

Having a set of tests run against both real impl and mock is a way to avoid this.

```
Env.Remove("targetEnv");
systemUnderTest = new RealEnv();
systemUnderTest.Get("targetEnv").ShouldBeNull;
systemUnderTest = new MockEnv();
systemUnderTest.Get("targetEnv").ShouldBeNull;
```

I say it is good for old API like this, because it is well documented. So it won't out of sync that bad. However, if it is an 3rd party API that is under rapid development. It's better not to assume the behavior.

Reply:
This is easily prevented by having your mock *always* conform to the documented (or even undocumented) contract of the interface being mocked.  If it doesn't conform to the contract, it's a bug in the mock (and if the test passes, it's a bug in your the code-under-test).

For example, a mock of `System.Environment.GetEnvironmentVariable` should never return empty string or throw an exception to indicate an environment variable was not set; it should return null as per the expected, documented behavior.  The code-under-test is expecting that contract, otherwise it is a bug in the code-under-test.

I agree that mocks aren't the be-all-end-all of unit testing and they certainly can't guarantee that the code-under-test is not buggy, but what they do well is enable testing of conditional behavior where the conditions are hard to replicate consistently in a test environment.